emscripten-simulated-jerry-api/jerry.c

```diff
@@ -0,0 +1,2002 @@
+                                 void **out_native_pointer_p, /**< [out] native pointer */
+                                 const jerry_object_native_info_t **out_native_info_p) /**< [out] the type info
+                                                                                        *   of the native pointer */
+{
+  if (!jerry_value_is_object (obj_val))
+  {
+    return false;
+  }
+  return (bool) EM_ASM_INT (
+    {
+      var value = __jerry.get ($0);
+      var internalProps = __jerry._jerryInternalPropsWeakMap.get (value);
+      var ptr = internalProps.nativePtr;
+      if (ptr === undefined)
+      {
+        return false;
+      }
+      if ($1)
+      {
+        Module.setValue ($1, ptr, '*');
+      }
+      if ($2)
+      {
+        Module.setValue ($2, internalProps.nativeInfo, '*');
+      }
+      return true;
+    }
+    , obj_val, out_native_pointer_p, out_native_info_p);
+} /* jerry_get_object_native_pointer */
+
+static bool __attribute__((used))
+_jerry_call_foreach_cb (jerry_object_property_foreach_t foreach_p,
+                        const jerry_value_t property_name,
+                        const jerry_value_t property_value,
+                        void *user_data_p)
+{
+  return foreach_p (property_name, property_value, user_data_p);
+} /* _jerry_call_foreach_cb */
+
+bool
+jerry_foreach_object_property (const jerry_value_t obj_val, /**< object value */
+                               jerry_object_property_foreach_t foreach_p, /**< foreach function */
+                               void *user_data_p) /**< user data for foreach function */
+{
+  return (bool) EM_ASM_INT (
+    {
+      var obj = __jerry.get ($0);
+      try
+      {
+        for (var propName in obj)
+        {
+          var propNameRef = __jerry.ref (propName);
+          var propValRef = __jerry.ref (obj[propName]);
+          var shouldContinue = Module.ccall (
+            '_jerry_call_foreach_cb',
+            'number',
+            ['number', 'number'],
+            [$1, propNameRef, propValRef, $2]);
+          if (!shouldContinue)
+          {
+            return true;
+          }
+        }
+      }
+      catch (e)
+      {
+        return false;
+      };
+      return true;
+    }
+    , obj_val, foreach_p, user_data_p
+  );
+} /* jerry_foreach_object_property */
+
+jerry_value_t
+jerry_resolve_or_reject_promise (jerry_value_t promise,
+                                 jerry_value_t argument,
+                                 bool is_resolve)
+{
+  if (!jerry_value_is_promise (promise))
+  {
+    return TYPE_ERROR_ARG;
+  }
+
+  return (jerry_value_t) EM_ASM_INT (
+    {
+      var p = __jerry.get ($0);
+      var arg = __jerry.get ($1);
+      var internalProps = __jerry._jerryInternalPropsWeakMap.get (p);
+      if ($2)
+      {
+        internalProps.promiseResolveFunc (arg);
+      }
+      else
+      {
+        internalProps.promiseRejectFunc (arg);
+      }
```

Comment:
There's a few bugs here:
- there is nothing returned here while a `jerry_value_t` should be returned
- if the `promiseResolveFunc` or `promiseRejectFunc`, it should be caught and do the error conversion we usually do (call `__jerry.setError` etc).

Reply:
fixed, thanks